lib/prioritized-attributes.js

```diff
@@ -0,0 +1,171 @@
+    const attrs = Object.create(null)
+
+    for (let [key, attr] of this.attributes) {
+      if (!(attr.destinations & dest)) {
+        continue
+      }
+
+      attrs[key] = typeof attr.value === 'string' && !attr.truncateExempt
+        ? byteUtils.truncate(attr.value, 255)
+        : attr.value
+    }
+
+    return attrs
+  }
+
+  has(key) {
+    this.attributes.has(key)
+  }
+
+  reset() {
+    this.attributes = new Map()
+  }
+
+  addAttribute(
+    destinations,
+    key,
+    value,
+    truncateExempt = false,
+    priority = ATTRIBUTE_PRIORITY.HIGH
+  ) {
+    const existingAttribute = this.attributes.get(key)
+
+    let hasDroppableAttributeKey = null
+    if (!existingAttribute && this.attributes.size === this.limit) {
+      hasDroppableAttributeKey = this._hasDroppableAttribute(priority)
+
+      if (!hasDroppableAttributeKey) {
+        logger.debug(
+          `Maximum number of custom attributes have been added.
+          Dropping attribute ${key} with ${value} type.`
+        )
+
+        return
+      }
+    }
+
+    if (existingAttribute && priority < existingAttribute.priority) {
+      logger.debug('incoming priority for \'%s\' is lower than existing, not updating.', key)
+      logger.trace(
+        '%s attribute retained value: %s, ignored value: %s',
+        key,
+        existingAttribute.value,
+        value
+      )
+      return
+    }
+
+    if (!isValidType(value)) {
+      logger.debug(
+        'Not adding attribute %s with %s value type. This is expected for undefined' +
+        'attributes and only an issue if an attribute is not expected to be undefined' +
+        'or not of the type expected.',
+        key,
+        typeof value
+      )
+      return
+    }
+
+    if (!this.isValidLength(key)) {
+      logger.warn(
+        'Length limit exceeded for attribute name, not adding: %s',
+        key
+      )
+      return
+    }
+
+
+    // Only set the attribute if at least one destination passed
+    const validDestinations = this.filter(destinations, key)
+    if (!validDestinations) {
+      return
+    }
+
+    if (hasDroppableAttributeKey) {
+      const droppableAttributeKey = this._lowPriorityAttributeNames.pop()
+      logger.trace(
+        'dropping existing lower priority attribute %s ' +
+        'to add higher priority attribute %s',
+        droppableAttributeKey,
+        key
+      )
+
+      this.attributes.delete(droppableAttributeKey)
+    }
+
+    this._set(validDestinations, key, value, truncateExempt, priority)
+  }
+
+  addAttributes(destinations, attrs) {
+    for (let key in attrs) {
+      if (properties.hasOwn(attrs, key)) {
+        this.addAttribute(destinations, key, attrs[key])
+      }
+    }
+  }
+
+  _hasDroppableAttribute(incomingPriority) {
+    // There will never be anything lower priority to drop
+    if (incomingPriority === ATTRIBUTE_PRIORITY.LOW) {
+      return
+    }
+
+    // Currently only support HIGH/LOW
+    if (!this._lowPriorityAttributeNames) {
```

Comment:
It looks like this is called every time `addAttribute` is called. Once `_lowPriorityAttributeNames` is created, it won't get updated again. I think the first time an attribute is added, it will create this array, but not add anything to it. After that, this array will never be updated, and the length will always be zero, so this will always return false. Is that right? Maybe the for loop below is supposed to be outside of the conditional above? Confused here..

Reply:
This will only be called when we are at the attribute maximum and there is no existing attribute to attempt to update. See the if statement on line 65. Once you are at the attribute maximum, it will never be possible for a new low-priority attribute to be added.

If we ever introduce a "medium" or whatever priority, then we'll have to tweak this optimization.

As such, the first time we look for low priority items we'll find all the low priority items that could ever exist. We'll then slowly pop them out until we know there are only high priority items.

I introduced this because there's no way to reverse-iterate the keys. If you want to do that, you first have to create an array from the collection, which iterates the collection anyways. A less optimal approach, but faster than doing that, is to just always iterate forward but at best that will be O(n) for each subsequent add. This only eats that cost once.

Reply:
oof, you did make me realize a gap where this could break tho. if a high priority attribute replaced a low priority attribute. i'll see if i want to fix this optimization or just ship with the slower method at first

Reply:
actually, i take it back. what i was worried about is not possible. but it does make me want a test.